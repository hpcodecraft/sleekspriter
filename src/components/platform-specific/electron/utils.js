import { writeFile } from "fs";
import { basename, dirname } from "path";

import config from "../../../config";
import { stateToFile } from "../../../utils";

export const save = (path, contents, errorHandler) => {
  writeFile(path, JSON.stringify(stateToFile(contents)), function(error) {
    if (error) {
      errorHandler(error);
    }
  });
};

export const fileInfoFromPath = path => {
  const fileName = basename(path, "." + config.fileExtension); // the file name without extension, e.g. "coin"
  const folder = dirname(path); // the complete absolute folder (same as path without the file name)

  return {
    fileName,
    folder,
  };
};

export const fileDefaultPath = (file, extension) => {
  const folder = file.folder || config.defaultFolder;
  const name = file.name || config.defaultName;
  const ext = extension || config.fileExtension;

  return `${folder}/${name}.${ext}`; // TODO: use platform-independent folder seperator
};
