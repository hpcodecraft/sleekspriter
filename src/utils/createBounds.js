import { selectionIsActive } from "./";

const createBounds = (size, selection) => {
  if (!selection || !selectionIsActive(selection)) {
    return {
      start: { x: 1, y: 1 },
      end: { x: size.width, y: size.height },
    };
  } else {
    return selection;
  }
};

export default createBounds;
