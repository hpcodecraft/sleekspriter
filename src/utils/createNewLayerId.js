import { maxBy } from "lodash";

const createNewLayerId = layers => {
  return maxBy(layers, "id").id + 1;
};

export default createNewLayerId;
