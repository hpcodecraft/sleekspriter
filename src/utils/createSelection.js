// this function takes two points and returns a selection object where the
// start always is the top left corner and
// end is the bottom right one
// this makes drawing the selection easier and ensures width calculations etc are always consistent

const createSelection = (p1, p2) => {
  return {
    start: {
      x: p1.x < p2.x ? p1.x : p2.x,
      y: p1.y < p2.y ? p1.y : p2.y,
    },
    end: {
      x: p1.x > p2.x ? p1.x : p2.x,
      y: p1.y > p2.y ? p1.y : p2.y,
    },
  };
};

export default createSelection;
