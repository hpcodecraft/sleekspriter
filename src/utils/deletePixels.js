const deletePixels = (allPixels, frame, layer, pixelsToDelete) => {

  Object.keys(pixelsToDelete).map(x => {
    Object.keys(pixelsToDelete[x]).map(y => {
      // console.log("delete! ", frame, layer, x, y);
      delete allPixels[frame][layer][x][y];
    });

    if(Object.keys(allPixels[frame][layer][x]).length === 0) {
      // console.log("delete! ", frame, layer, x);
      delete allPixels[frame][layer][x];
    }
  });

  if(Object.keys(allPixels[frame][layer]).length === 0) {
    // console.log("delete! ", frame, layer);
    delete allPixels[frame][layer];
  }
  if(Object.keys(allPixels[frame]).length === 0) {
    // console.log("delete! ", frame);
    delete allPixels[frame];
  }

  return allPixels;
};

export default deletePixels;
