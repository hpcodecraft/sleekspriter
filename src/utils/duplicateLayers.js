const duplicateLayers = (layers, targetFrame, nextLayerId) => {
  const newLayers = [];
  for (let i = 0; i < layers.length; i++) {
    const newId = nextLayerId + i;
    const newLayer = Object.assign({}, layers[i]);
    newLayer.id = newId;
    newLayer.frame = targetFrame;
    newLayer.name = `${newLayer.name} copy`;
    newLayers.push(newLayer);
  }

  return newLayers;
};

export default duplicateLayers;
