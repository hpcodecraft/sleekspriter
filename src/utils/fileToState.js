// converts JSON from file format to state-ready JSON
// TODO: add animations

const fileToState = file => {
  let { frames, layers, pixels, size } = file;

  layers = layers.map(layerToState);

  const frameDict = {};
  layers.forEach(layer => {
    frameDict[layer.id] = layer.frame;
  });

  const mappedPixels = {};
  pixels.forEach(p => {
    const pixel = {
      frame: frameDict[p[0]],
      layer: p[0],
      x: p[1],
      y: p[2],
      r: p[3],
      g: p[4],
      b: p[5],
      a: p[6],
    };

    if (!mappedPixels[pixel.frame]) mappedPixels[pixel.frame] = {};
    if (!mappedPixels[pixel.frame][pixel.layer])
      mappedPixels[pixel.frame][pixel.layer] = {};
    if (!mappedPixels[pixel.frame][pixel.layer][pixel.x])
      mappedPixels[pixel.frame][pixel.layer][pixel.x] = {};
    mappedPixels[pixel.frame][pixel.layer][pixel.x][pixel.y] = pixel;
  });

  return {
    frames: {
      x: frames[0],
      y: frames[1],
    },
    layers,
    pixels: mappedPixels,
    size: {
      width: size[0],
      height: size[1],
    },
  };
};

const layerToState = layer => ({
  id: layer[0],
  frame: layer[1],
  name: layer[2],
  z: layer[3],
  opacity: layer[4],
  visible: !!layer[5],
});

export default fileToState;
