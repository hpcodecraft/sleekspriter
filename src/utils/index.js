export capitalizeFirstLetter from "./capitalizeFirstLetter";
export createBounds from "./createBounds";
export createNewLayerId from "./createNewLayerId";
export deletePixels from "./deletePixels";
export duplicateLayers from "./duplicateLayers";
export flattenPixels from "./flattenPixels";
export getPivot from "./getPivot";
export getPixelsInFrame from "./getPixelsInFrame";
export getPixelsInScope from "./getPixelsInScope";
export fileToState from "./fileToState";
export insideBounds from "./insideBounds";
export mergeLayerPixels from "./mergeLayerPixels";
export selectionIsActive from "./selectionIsActive";
export stateToFile from "./stateToFile";
export t from "./translate";
