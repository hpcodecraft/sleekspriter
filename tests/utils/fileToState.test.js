jest.unmock("utils");

import { fileToState } from "utils";

describe("fileToState", () => {
  it("maps JSON from file format to redux state", () => {
    const file = {
      size: [20, 20],
      frames: [7, 3],
      layers: [
        [4, 1, "lights", 2, 100, 1],
        [3, 1, "shadows", 1, 100, 1],
        [20, 20, "layer 20", 0, 100, 1],
        [19, 19, "layer 19", 0, 100, 1],
        [18, 18, "layer 18", 0, 100, 1],
        [17, 17, "layer 17", 0, 100, 1],
        [16, 16, "layer 16", 0, 100, 1],
        [15, 15, "layer 15", 0, 100, 1],
        [14, 13, "layer 14", 0, 100, 1],
        [12, 11, "layer 12", 0, 100, 1],
        [11, 10, "layer 11", 0, 100, 1],
        [10, 9, "layer 10", 0, 100, 1],
        [8, 6, "layer 8", 0, 100, 1],
        [7, 5, "layer 7", 0, 100, 1],
        [6, 4, "layer 6", 0, 100, 1],
        [5, 3, "layer 5", 0, 100, 1],
        [2, 2, "layer 2", 0, 100, 1],
        [1, 1, "shape", 0, 100, 1],
        [21, 8, "lights", 2, 100, 1],
        [22, 8, "shadows", 1, 100, 1],
        [23, 8, "shape", 0, 100, 1],
        [24, 7, "Layer 24", 0, 100, 1],
        [25, 14, "Layer 25", 0, 100, 1],
        [26, 21, "Layer 26", 0, 100, 1],
        [27, 12, "layer 11", 0, 100, 1]
      ],
      animations: [
        [2, "Animation 2", 10, [1, 7]],
        [1, "rotate", 10, [8, 9, 10, 11, 10, 9]]
      ],
      pixels: [
        [4, 13, 5, 254, 242, 128, 1],
        [4, 12, 5, 254, 242, 128, 1],
        [4, 11, 5, 254, 242, 128, 1]
      ]
    };

    expect(fileToState(file)).toMatchSnapshot();
  });
});
