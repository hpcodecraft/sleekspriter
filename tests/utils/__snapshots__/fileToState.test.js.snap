// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`maps JSON from file format to redux state 1`] = `
Object {
  "frames": Object {
    "x": 7,
    "y": 3,
  },
  "layers": Array [
    Object {
      "frame": 1,
      "id": 4,
      "name": "lights",
      "opacity": 100,
      "visible": true,
      "z": 2,
    },
    Object {
      "frame": 1,
      "id": 3,
      "name": "shadows",
      "opacity": 100,
      "visible": true,
      "z": 1,
    },
    Object {
      "frame": 20,
      "id": 20,
      "name": "layer 20",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 19,
      "id": 19,
      "name": "layer 19",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 18,
      "id": 18,
      "name": "layer 18",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 17,
      "id": 17,
      "name": "layer 17",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 16,
      "id": 16,
      "name": "layer 16",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 15,
      "id": 15,
      "name": "layer 15",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 13,
      "id": 14,
      "name": "layer 14",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 11,
      "id": 12,
      "name": "layer 12",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 10,
      "id": 11,
      "name": "layer 11",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 9,
      "id": 10,
      "name": "layer 10",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 6,
      "id": 8,
      "name": "layer 8",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 5,
      "id": 7,
      "name": "layer 7",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 4,
      "id": 6,
      "name": "layer 6",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 3,
      "id": 5,
      "name": "layer 5",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 2,
      "id": 2,
      "name": "layer 2",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 1,
      "id": 1,
      "name": "shape",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 8,
      "id": 21,
      "name": "lights",
      "opacity": 100,
      "visible": true,
      "z": 2,
    },
    Object {
      "frame": 8,
      "id": 22,
      "name": "shadows",
      "opacity": 100,
      "visible": true,
      "z": 1,
    },
    Object {
      "frame": 8,
      "id": 23,
      "name": "shape",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 7,
      "id": 24,
      "name": "Layer 24",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 14,
      "id": 25,
      "name": "Layer 25",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 21,
      "id": 26,
      "name": "Layer 26",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
    Object {
      "frame": 12,
      "id": 27,
      "name": "layer 11",
      "opacity": 100,
      "visible": true,
      "z": 0,
    },
  ],
  "pixels": Object {
    "1": Object {
      "4": Object {
        "11": Object {
          "5": Object {
            "a": 1,
            "b": 128,
            "frame": 1,
            "g": 242,
            "layer": 4,
            "r": 254,
            "x": 11,
            "y": 5,
          },
        },
        "12": Object {
          "5": Object {
            "a": 1,
            "b": 128,
            "frame": 1,
            "g": 242,
            "layer": 4,
            "r": 254,
            "x": 12,
            "y": 5,
          },
        },
        "13": Object {
          "5": Object {
            "a": 1,
            "b": 128,
            "frame": 1,
            "g": 242,
            "layer": 4,
            "r": 254,
            "x": 13,
            "y": 5,
          },
        },
      },
    },
  },
  "size": Object {
    "height": 20,
    "width": 20,
  },
}
`;
